pipeline/load.py

```diff
@@ -0,0 +1,314 @@
+    """Inserts data relating to a track (belonging to an album) sale."""
+
+    if not all(
+        field in track_sale and track_sale[field] for field in REQUIRED_FIELDS_TRACK
+    ):
+        logging.warning(
+            "Skipping track sale due to missing required fields: %s", track_sale
+        )
+        return
+
+    country_id = get_or_insert_country(cursor, track_sale["country"])
+    artist_id = get_or_insert_artist(
+        cursor, track_sale["artist_name"], track_sale["artist_url"]
+    )
+    album_id = get_or_insert_album(
+        cursor, track_sale["album_title"], artist_id, track_sale["album_url"]
+    )
+
+    track_id = get_or_insert_track_or_single(
+        cursor, track_sale["item_description"], artist_id, track_sale["url"], album_id
+    )
+
+    for tag in track_sale["track_tags"]:
+        get_or_insert_tags(cursor, tag, track_id=track_id)
+
+    insert_album_or_track_purchase(
+        cursor,
+        track_sale["utc_date"],
+        track_sale["amount_paid_usd"],
+        country_id,
+        track_id=track_id,
+    )
+
+
+def insert_single_sale(cursor: DBCursor, single_sale: Dict[str, Any]) -> None:
+    """Inserts data relating to a single sale."""
+
+    if not all(
+        field in single_sale and single_sale[field] for field in REQUIRED_FIELDS_SINGLE
+    ):
+        logging.warning(
+            "Skipping single sale due to missing required fields: %s, single_sale"
+        )
+        return
+
+    country_id = get_or_insert_country(cursor, single_sale["country"])
+    artist_id = get_or_insert_artist(
+        cursor, single_sale["artist_name"], single_sale["artist_url"]
+    )
+    track_id = get_or_insert_track_or_single(
+        cursor, single_sale["item_description"], artist_id, single_sale["url"]
+    )
+
+    for tag in single_sale["track_tags"]:
+        get_or_insert_tags(cursor, tag, track_id=track_id)
+
+    insert_album_or_track_purchase(
+        cursor,
+        single_sale["utc_date"],
+        single_sale["amount_paid_usd"],
+        country_id,
+        track_id=track_id,
+    )
+
+
+def load_sales_data(sales_data: List[Dict[str, Any]]) -> None:
+    """Loads sales data into the database."""
+
+    try:
+        connection = get_connection()
+        with connection.cursor() as cursor:
+            for sale in sales_data:
+                logging.info("Processing sale: %s", sale)
+                if sale["item_type"] == "a":
+                    logging.info("Inserting album sale")
+                    insert_album_sale(cursor, sale)
+                elif sale["item_type"] == "t":
+                    if sale.get("album_title"):
+                        logging.info("Inserting track sale")
+                        insert_track_sale(cursor, sale)
+                else:
+                    logging.info("Inserting single sale")
+                    insert_single_sale(cursor, sale)
+        connection.commit()
+    except Exception as e:
+        logging.error("An error occurred: %s", e)
```

Comment:
Could you use a rollback here?

Reply:
Looking at the [docs](https://www.psycopg.org/docs/connection.html#:~:text=Closing%20a%20connection%20without%20committing%20the%20changes%20first%20will%20cause%20an%20implicit%20rollback%20to%20be%20performed), in this case I think it would rollback automatically when an error occurs, so it should be okay as it is.